proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy

```diff
@@ -462,106 +558,194 @@ crunch perform_page_table_invocation, perform_page_directory_invocation, perform
-  apply (forward_inv_step wp: check_cap_inv)+
-  by (wpsimp wp: check_cap_inv hoare_drop_imps thread_set_hyp_refs_trivial thread_set_valid_cur_vcpu)
+  by (wpsimp wp: hoare_drop_imps check_cap_inv)
+     (wpsimp wp: hoare_drop_imps | wpsimp wp: check_cap_inv thread_set_valid_cur_vcpu_unchanged | wp_pre, wps)+
 
-crunch set_domain
-  for arch_state[wp]: "\<lambda>s. P (arch_state s)"
-  and arch_tcb_at[wp]: "arch_tcb_at P t"
-  and cur_thread[wp]: "\<lambda>s. P (cur_thread s)"
-  and valid_cur_vcpu[wp]: valid_cur_vcpu
-  (wp: valid_cur_vcpu_lift_weak thread_set_no_change_tcb_pred crunch_wps)
+lemma thread_set_domain_valid_cur_vcpu[wp]:
+  "\<lbrace>\<lambda>s. (valid_cur_vcpu s \<or> (tptr = cur_thread s \<and> cur_domain s \<noteq> new_dom))
+        \<and> in_cur_domain (cur_thread s) s\<rbrace>
+   thread_set_domain tptr new_dom
+   \<lbrace>\<lambda>_. valid_cur_vcpu\<rbrace>"
+  unfolding thread_set_domain_def thread_set_def
+  apply (wpsimp wp: set_object_wp)
+  by (auto simp: valid_cur_vcpu_def active_cur_vcpu_of_def in_cur_domain_def pred_tcb_at_def
+                 obj_at_def get_tcb_def etcb_at_def etcbs_of'_def)
 
-crunch invoke_domain
-  for valid_cur_vcpu[wp]: valid_cur_vcpu
+lemma set_domain_valid_cur_vcpu[wp]:
+  "\<lbrace>\<lambda>s. (valid_cur_vcpu s \<or> (tptr = cur_thread s \<and> cur_domain s \<noteq> new_dom))
+        \<and> in_cur_domain (cur_thread s) s\<rbrace>
+   set_domain tptr new_dom
+   \<lbrace>\<lambda>_. valid_cur_vcpu\<rbrace>"
+  unfolding set_domain_def
+  by (wpsimp wp: hoare_vcg_disj_lift | wps)+
+
+lemma vcpu_flush_valid_cur_vcpu_not_active:
+  "\<lbrace>\<lambda>s. valid_cur_vcpu s \<and> active_cur_vcpu_of s = None \<and> in_cur_domain (cur_thread s) s\<rbrace>
+   vcpu_flush
+   \<lbrace>\<lambda>_ s. valid_cur_vcpu s\<rbrace>"
+  unfolding vcpu_flush_def
+  apply wpsimp
+  by (auto simp: valid_cur_vcpu_def)
+
+crunch arch_prepare_set_domain
+  for cur_thread[wp]: "\<lambda>s. P (cur_thread s)"
+  and cur_domain[wp]: "\<lambda>s. P (cur_domain s)"
+  and etcbs_of[wp]: "\<lambda>s. P (etcbs_of s)"
+  (wp: crunch_wps simp: crunch_simps)
+
+\<comment> \<open>FIXME: move earlier, along with sym_refs_tcb_vcpu and sym_refs_vcpu_tcb from Refine\<close>
+lemma sym_refs_kheap_tcb_vcpu:
+  "\<lbrakk>kheap s tptr = Some (TCB tcb); tcb_vcpu (tcb_arch tcb) = Some vcpuptr; sym_refs (state_hyp_refs_of s)\<rbrakk>
+   \<Longrightarrow> \<exists>vcpu. kheap s vcpuptr = Some (ArchObj (VCPU vcpu)) \<and> vcpu_tcb vcpu = Some tptr"
+  apply (drule_tac x=tptr in sym_refsD[rotated])
+   apply (fastforce simp: state_hyp_refs_of_def)
+  apply (auto simp: state_hyp_refs_of_def hyp_refs_of_def refs_of_a_def tcb_vcpu_refs_def vcpu_tcb_refs_def
+             split: option.splits kernel_object.splits arch_kernel_obj.splits)
+  done
+
+lemma arch_prepare_set_domain_valid_cur_vcpu[wp]:
+  "\<lbrace>\<lambda>s. valid_cur_vcpu s \<and> sym_refs (state_hyp_refs_of s) \<and> in_cur_domain (cur_thread s) s\<rbrace>
```

Comment:
almost tempting to have an abbreviation for `(%s. in_cur_domain (cur_thread s) s)` so we can keep using `and`. probably not worth the bother